homeassistant/components/input_datetime.py

```diff
@@ -0,0 +1,214 @@
+    def async_added_to_hass(self):
+        """Run when entity about to be added."""
+        if self._current_datetime is not None:
+            return
+
+        old_state = yield from async_get_last_state(self.hass, self.entity_id)
+        if old_state is not None:
+            restore_val = dt_util.parse_datetime(old_state.state)
+        else:
+            restore_val = dt_util.now()
+
+        print("Type is: " + str(type(restore_val)))
+        if not self._has_date:
+            self._current_datetime = restore_val.time()
+        elif not self._has_time:
+            self._current_datetime = restore_val.date()
+        else:
+            self._current_datetime = restore_val
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the select input."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the state of the component."""
+        if self._current_datetime is None:
+            return None
+
+        print("Returning type: " + str(type(self._current_datetime)))
+        return self._current_datetime
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        attrs = {
+            'has_date': self._has_date,
+            'has_time': self._has_time,
+        }
+
+        if self._has_date and self._current_datetime is not None:
+            attrs['year'] = self._current_datetime.year
+            attrs['month'] = self._current_datetime.month
+            attrs['day'] = self._current_datetime.day
+        if self._has_time and self._current_datetime is not None:
+            attrs['hour'] = self._current_datetime.hour
+            attrs['minute'] = self._current_datetime.minute
+            attrs['second'] = self._current_datetime.second
+
+        return attrs
+
+    @asyncio.coroutine
+    def async_set_datetime(self, date_val, time_val):
+        """Set a new date / time."""
+        if not self._has_date:
+            if time_val is None:
+                _LOGGER.warning('"None" passed as time.')
+                return
+            self._current_datetime = time_val
+        elif not self._has_time:
+            if date_val is None:
+                _LOGGER.warning('"None" passed as date.')
+                return
+            self._current_datetime = date_val
+        else:
+            if time_val is None:
+                _LOGGER.warning('"None" passed as time.')
+                return
+            if date_val is None:
+                _LOGGER.warning('"None" passed as date.')
+                return
+            self._current_datetime = datetime.datetime.combine(date_val,
+                                                               time_val)
+
+        yield from self.async_update_ha_state()
```

Comment:
move that into handler function

Reply:
I don't understand: isn't `async_set_datetime` already the handler function? It directly sets the internal state (`self._current_datetime`), no further handler is called, or am I missing something? Or should I add another handler function, schedule a call to this handler function at the end of `async_set_datetime`, and then put `self.async_update_ha_state()` into it? 

I need to call `self.async_update_ha_state()` every time I modify the internal state of an entity, right? Such that the state machine and the frontend update their state?